Source/Core/Core/IOS/USB/Emulated/Infinity.h

```diff
@@ -0,0 +1,125 @@
+  ~InfinityUSB() override;
+  DeviceDescriptor GetDeviceDescriptor() const override;
+  std::vector<ConfigDescriptor> GetConfigurations() const override;
+  std::vector<InterfaceDescriptor> GetInterfaces(u8 config) const override;
+  std::vector<EndpointDescriptor> GetEndpoints(u8 config, u8 interface, u8 alt) const override;
+  bool Attach() override;
+  bool AttachAndChangeInterface(u8 interface) override;
+  int CancelTransfer(u8 endpoint) override;
+  int ChangeInterface(u8 interface) override;
+  int GetNumberOfAltSettings(u8 interface) override;
+  int SetAltSetting(u8 alt_setting) override;
+  int SubmitTransfer(std::unique_ptr<CtrlMessage> message) override;
+  int SubmitTransfer(std::unique_ptr<BulkMessage> message) override;
+  int SubmitTransfer(std::unique_ptr<IntrMessage> message) override;
+  int SubmitTransfer(std::unique_ptr<IsoMessage> message) override;
+
+private:
+  void ScheduleTransfer(std::unique_ptr<TransferCommand> command, const std::array<u8, 32>& data,
+                        s32 expected_count, u64 expected_time_us);
+
+  Kernel& m_ios;
+  u16 m_vid = 0;
+  u16 m_pid = 0;
+  u8 m_active_interface = 0;
+  bool m_device_attached = false;
+  DeviceDescriptor m_device_descriptor;
+  std::vector<ConfigDescriptor> m_config_descriptor;
+  std::vector<InterfaceDescriptor> m_interface_descriptor;
+  std::vector<EndpointDescriptor> m_endpoint_descriptor;
+  std::queue<std::array<u8, 32>> m_queries;
+  std::queue<std::unique_ptr<IntrMessage>> m_response_list;
+};
+
+class InfinityBase final
+{
+public:
+  bool HasFigureBeenAddedRemoved() const;
+  std::array<u8, 32> PopAddedRemovedResponse();
+  void GetBlankResponse(u8 sequence, u8* reply_buf);
+  void GetPresentFigures(u8 sequence, u8* reply_buf);
+  void GetFigureIdentifier(u8 fig_num, u8 sequence, u8* reply_buf);
+  void QueryBlock(u8 fig_num, u8 block, u8* reply_buf, u8 sequence);
+  void WriteBlock(u8 fig_num, u8 block, const u8* to_write_buf, u8* reply_buf, u8 sequence);
+  void RemoveFigure(u8 position);
+  // Returns Infinity Figure name based on data from in_file param
+  std::string LoadFigure(u8* buf, File::IOFile in_file, u8 position);
+  bool CreateFigure(const std::string& file_path, u16 character);
+  static const std::array<std::pair<const char*, const InfinityFigureInfo>, 104>& GetFigureList();
+  std::string FindFigure(u16 character) const;
+  void GenerateSeed(u32 seed);
+  u32 GetNext();
+  u64 Scramble(u32 num_to_scramble, u32 garbage);
+  u32 Descramble(u64 num_to_descramble);
+  u8 GenerateChecksum(const u8* data, int num_of_bytes) const;
+
+protected:
+  std::mutex m_infinity_mutex;
+  std::array<InfinityFigure, 7> m_figures;
+
+private:
+  InfinityFigure& GetFigureByOrder(u8 order_added);
+  std::array<u8, 16> GenerateInfinityFigureKey(std::vector<u8> sha1_data);
+  std::array<u8, 16> GenerateBlankFigureData(u16 figure_num);
+
+  // These 4 variables are state variables used during the seeding of, and use of the random number
```

Comment:
The seeding of what?

Reply:
Whoops, poor english in the comment here, should read as the seeding and use of the random number generator